In [ ]:
import pandas as pd
import numpy as np

In [ ]:
from google.colab import drive
drive.mount('/content/drive')

Drive already mounted at /content/drive; to attempt to forcibly remount, call drive.mount("/content/drive", force_remount=True).


In [ ]:
path='/content/drive/MyDrive/Hamoye Dataset/energydata_complete.csv'
df=pd.read_csv(path)
df.head(1)

date  Appliances  lights  ...  Tdewpoint        rv1        rv2
0  2016-01-11 17:00:00          60      30  ...        5.3  13.275433  13.275433

[1 rows x 29 columns]

# Question 12. 

##  From the dataset, fit a linear model on the relationship between the temperature in the living room in Celsius (x = T2) and the temperature outside the building (y = T6). What is the R^2 value in two d.p.?

In [ ]:
from sklearn import linear_model
model = linear_model.LinearRegression()
train_x = np.asanyarray(df[['T2']])
train_y = np.asanyarray(df[['T6']])
model.fit (train_x, train_y)
#finding the coefficients
print ('Coefficients: ', model.coef_)
print ('Intercept: ',model.intercept_)

Coefficients:  [[2.22506139]]
Intercept:  [-37.34952264]


Training and Testing is done on the same dataset.

In [ ]:
from sklearn.metrics import r2_score

test_x = np.asanyarray(df[['T2']])
test_y = np.asanyarray(df[['T6']])
predict_y = model.predict(test_x)
#metrics

print("Mean absolute error: %.2f" % np.mean(np.absolute(predict_y - test_y)))
print("Residual sum of squares (MSE): %.2f" % np.mean((predict_y - test_y) ** 2))
print("R2-score: %.2f" % r2_score(test_y , predict_y) )

Mean absolute error: 2.84
Residual sum of squares (MSE): 13.28
R2-score: 0.64


## Answer R2 score equals to 0.64

In [ ]:
new_df=df.drop(columns=['date', 'lights'],axis=1)

In [ ]:
new_df.head(1)

Appliances     T1       RH_1  ...  Tdewpoint        rv1        rv2
0          60  19.89  47.596667  ...        5.3  13.275433  13.275433

[1 rows x 27 columns]

In [ ]:
df.head(1)

date  Appliances  lights  ...  Tdewpoint        rv1        rv2
0  2016-01-11 17:00:00          60      30  ...        5.3  13.275433  13.275433

[1 rows x 29 columns]

In [ ]:
from sklearn.preprocessing import MinMaxScaler
scaler= MinMaxScaler()
normalised_df=pd.DataFrame(scaler.fit_transform(new_df),columns=new_df.columns)
df_features=normalised_df.drop(['Appliances'],axis=1)
df_target=normalised_df['Appliances']

In [ ]:
normalised_df.head()

Appliances       T1      RH_1  ...  Tdewpoint       rv1       rv2
0    0.046729  0.32735  0.566187  ...   0.538462  0.265449  0.265449
1    0.046729  0.32735  0.541326  ...   0.533937  0.372083  0.372083
2    0.037383  0.32735  0.530502  ...   0.529412  0.572848  0.572848
3    0.037383  0.32735  0.524080  ...   0.524887  0.908261  0.908261
4    0.046729  0.32735  0.531419  ...   0.520362  0.201611  0.201611

[5 rows x 27 columns]

In [ ]:
from sklearn.model_selection import train_test_split
x_train,x_test,y_train,y_test= train_test_split(df_features,df_target, test_size=0.3, random_state=42)

In [ ]:
print('the shape of x_train:{}'.format(x_train.shape))
print('the shape of x_test:{}'.format(x_test.shape))
print('the shape of y_train:{}'.format(y_train.shape))
print('the shape of y_test:{}'.format(y_test.shape))

the shape of x_train:(13814, 26)
the shape of x_test:(5921, 26)
the shape of y_train:(13814,)
the shape of y_test:(5921,)


In [ ]:
from sklearn.linear_model import LinearRegression
linear_model=LinearRegression()

In [ ]:
#fit the model to the training dataset
linear_model.fit(x_train, y_train)

LinearRegression()

In [ ]:
y_predict=linear_model.predict(x_test)

# Question 13 

## What is the Mean Absolute Error (in two decimal places)?

In [ ]:
# mean absolute error or the average error MAE
from sklearn.metrics import mean_absolute_error
mae= mean_absolute_error(y_test, y_predict)
round(mae, 2)

0.05

## Question 14
## What is the Residual Sum of Squares (in two decimal places)?

In [ ]:
rss = np.sum(np.square(y_test - y_predict))
round(rss, 2)

45.35

# Question 15

## What is the Root Mean Squared Error (in three decimal places)?

In [ ]:
from sklearn.metrics import mean_squared_error
rmse = np.sqrt(mean_squared_error(y_test,y_predict))
round(rmse, 3)

0.088

## Question 16
## What is the Coefficient of Determination (in two decimal places)?

In [ ]:
from sklearn.metrics import r2_score
r2_score = r2_score(y_test,y_predict)
round(r2_score, 2)

0.15

## Question 17
## Obtain the feature weights from your linear model above. Which features have the lowest and highest weights respectively?

In [ ]:
def get_weights_energy(model, feat, col_name):
#this function returns the weight of every feature
  weights = pd.Series(model.coef_, feat.columns).sort_values()
  weights_energy = pd.DataFrame(weights).reset_index()
  weights_energy.columns = ['Features', col_name]
  weights_energy[col_name].round(3)
  return weights_energy

In [ ]:
linear_model_weights = get_weights_energy(linear_model, x_train, 'Linear_Model_Weight')


In [ ]:
linear_model_weights

Features  Linear_Model_Weight
0          RH_2            -0.456698
1         T_out            -0.321860
2            T2            -0.236178
3            T9            -0.189941
4          RH_8            -0.157595
5        RH_out            -0.077671
6          RH_7            -0.044614
7          RH_9            -0.039800
8            T5            -0.015657
9            T1            -0.003281
10          rv1             0.000770
11          rv2             0.000770
12  Press_mm_hg             0.006839
13           T7             0.010319
14   Visibility             0.012307
15         RH_5             0.016006
16         RH_4             0.026386
17           T4             0.028981
18    Windspeed             0.029183
19         RH_6             0.038049
20         RH_3             0.096048
21           T8             0.101995
22    Tdewpoint             0.117758
23           T6             0.236425
24           T3             0.290627
25         RH_1             0.553547

Since the values of the coefficient is sorted from lowest to highest.

Answer: the lowest coefficient RH_2	with -0.456698, and the highest coefficient is RH_1	is 0.553547.

# Question 18 
## Train a ridge regression model with an alpha value of 0.4. Is there any change to the root mean squared error (RMSE) when evaluated on the test set?

## Ridge Regression

In [ ]:
from sklearn.linear_model import Ridge
ridge_reg = Ridge(alpha=0.4)
ridge_reg.fit(x_train, y_train)

Ridge(alpha=0.4)

In [ ]:
y_predicts=ridge_reg.predict(x_test)
y_predicts

array([0.03321872, 0.24043824, 0.03461337, ..., 0.06872351, 0.10025536,
       0.05851175])

In [ ]:
from sklearn.metrics import mean_squared_error
rmse = np.sqrt(mean_squared_error(y_test,y_predicts))
round(rmse, 3)

0.088

We can notice that the RMSE using linear Regression and Rigde Regression both yield the same results of 0.088

Answer: No

## Question 19
## Train a lasso regression model with an alpha value of 0.001 and obtain the new feature weights with it. How many of the features have non-zero feature weights?

## Lasso Regression

In [ ]:
from sklearn.linear_model import Lasso
lasso_reg = Lasso(alpha=0.001)
lasso_reg.fit(x_train, y_train)

Lasso(alpha=0.001)

In [ ]:
def get_weights_energy(model, feat, col_name):
#this function returns the weight of every feature
  weights = pd.Series(model.coef_, feat.columns).sort_values()
  weights_energy = pd.DataFrame(weights).reset_index()
  weights_energy.columns = ['Features', col_name]
  weights_energy[col_name].round(3)
  return weights_energy

In [ ]:
lasso_reg_weights = get_weights_energy(lasso_reg, x_train, 'lasso_reg_Weight')

In [ ]:
lasso_reg_weights

Features  lasso_reg_Weight
0        RH_out         -0.049557
1          RH_8         -0.000110
2            T1          0.000000
3     Tdewpoint          0.000000
4    Visibility          0.000000
5   Press_mm_hg         -0.000000
6         T_out          0.000000
7          RH_9         -0.000000
8            T9         -0.000000
9            T8          0.000000
10         RH_7         -0.000000
11          rv1         -0.000000
12           T7         -0.000000
13           T6          0.000000
14         RH_5          0.000000
15           T5         -0.000000
16         RH_4          0.000000
17           T4         -0.000000
18         RH_3          0.000000
19           T3          0.000000
20         RH_2         -0.000000
21           T2          0.000000
22         RH_6         -0.000000
23          rv2         -0.000000
24    Windspeed          0.002912
25         RH_1          0.017880

The features having non-zero feature weights are:

1. RH_out with coefficient -0.049557
2. RH_8 with coefficient -0.000110
3. Windspeed with coefficient 0.002912
4. RH_1 with coefficient 0.017880

Answer: 4

## Question 20.
## What is the new RMSE with the lasso regression? (Answer should be in three (3) decimal places)

In [ ]:
y_predictss=lasso_reg.predict(x_test)
y_predictss

array([0.07370267, 0.08143458, 0.07716072, ..., 0.07792848, 0.09034412,
       0.08359255])

In [ ]:
from sklearn.metrics import mean_squared_error
rmse = np.sqrt(mean_squared_error(y_test,y_predictss))
round(rmse, 3)

0.094

Answer: The root mean squared error (RMSE) is equal to 0.094